gudpy/gui/widgets/core/worker.py

```diff
@@ -1,252 +1,235 @@
-            purge_det = os.path.join(sys._MEIPASS, f"{self.PROCESS}{SUFFIX}")
-        else:
-            purge_det = utils.resolve(
-                os.path.join(
-                    config.__rootdir__, "bin"
-                ), f"{self.PROCESS}{SUFFIX}"
-            )
-        if not os.path.exists(purge_det):
-            self.errorOccured.emit("MISSING_BINARY")
-            return
+        self.name = ""
+        self.output = ""
 
-        with tempfile.TemporaryDirectory() as tmp:
-            self.gudrunFile.setGudrunDir(tmp)
-            self.gudrunFile.purgeFile.write_out(os.path.join(
-                self.gudrunFile.instrument.GudrunInputFileDir,
-                f"{self.PROCESS}.dat"
-            ))
-
-            with subprocess.Popen(
-                [purge_det, f"{self.PROCESS}.dat"], cwd=tmp,
-                stdout=subprocess.PIPE,
-                stderr=subprocess.STDOUT
-            ) as purge:
-                for line in purge.stdout:
-                    self.outputChanged.emit(line.decode("utf8").rstrip("\n"))
-                if purge.stderr:
-                    self.errorOccured.emit(
-                        purge.stderr.decode("utf8").rstrip("\n"))
-                    return
-
-            self.gudrunFile.purgeOutput = (
-                self.gudrunFile.purgeFile.organiseOutput()
-            )
+    def _outputChanged(self, output):
+        self.output += output
+        self.outputChanged.emit(output)
+        self.progressChanged.emit(self._progressChanged(), self.name)
 
-        self.gudrunFile.setGudrunDir(
-            self.gudrunFile.projectDir)
 
-        self.finished.emit(1)
+class PurgeWorker(Worker, gudpy.Purge):
+    def __init__(self, purgeFile: PurgeFile, gudrunFile: GudrunFile):
+        super().__init__()
+        self.name = "Purge"
+        self.purgeFile = purgeFile
+        self.detectors = None
+        self.dataFileType = gudrunFile.instrument.dataFileType
+        self.dataFiles = [gudrunFile.instrument.groupFileName]
+
+        self.appendDfs(gudrunFile.normalisation.dataFiles[0])
+        self.appendDfs(gudrunFile.normalisation.dataFilesBg[0])
+        self.appendDfs([df for sb in gudrunFile.sampleBackgrounds
+                        for df in sb.dataFiles])
+        if not purgeFile.excludeSampleAndCan:
+            self.appendDfs([df for sb in gudrunFile.sampleBackgrounds
+                            for s in sb.samples for df in s.dataFiles
+                            if s.runThisSample])
+            self.appendDfs([df for sb in gudrunFile.sampleBackgrounds
+                            for s in sb.samples for c in s.containers
+                            for df in c.dataFiles if s.runThisSample])
+
+    def _progressChanged(self):
+        stepSize = math.ceil(100 / len(self.dataFiles))
+        progress = 0
+        for df in self.dataFiles:
+            if df in self.output:
+                progress += stepSize
+        return progress
+
+    def appendDfs(self, dfs):
```

Comment:
Function name is a bit obtuse! Could we go for something more like:
```suggestion
    def appendDataFiles(self, dfs):
```